<div style="background-color:#f9f9f9; padding:10px; border:1px solid #ddd; border-radius:8px;">

<h1 style='text-align:center;color:#2c3e50;font-family:Verdana;'><strong>Big Data Science - Fall 2023</strong></h1>

<h2 style='text-align:center;color:#3498db;font-family:Verdana;'>Assignment 2: Large Language Models on Protein Sequences</h2>

<hr style="border-top: 2px solid #3498db; margin-bottom: 20px; margin-top: 20px;">

<h3 style='color:#2c3e50;font-family:Verdana;'>Milestones:</h3>

<ol type="1" style='font-family:Verdana;'>
    <li><span style='color:#e74c3c;'>Understanding and Implementing Attention Mechanism and BERT Architecture</span></li>
    <li><span style='color:#27ae60;'>Fine-tuning Model on 2 Biological Tasks</span></li>
    <li><span style='color:#3498db;'>Evaluating Models' Performance with or without pretraining</span></li>
</ol>

</div>


<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Submission Deadline: Friday, December 8 at 8:59 P.M.</strong></h3>

<p style='color:#2c3e50;'><strong>Notes before you embark on this assignment journey:</strong></p>

<ul style='color:#2c3e50;'>
    <li><strong>Avoid using built-in libraries</strong> for the core algorithms, strive to build from scratch to strengthen your understanding. If you are allowed to use built-in libraries, you will be notified in the respective sections.
    </li>
    <li><strong>Collaboration is key</strong> to successful learning, feel free to discuss problems with your peers but ensure your submissions are individual efforts.</li>
    <li><strong>The early bird catches the worm!</strong> Start early to allow ample time for exploring, implementing, and seeking help if needed.</li>
    <li><strong>Ensure your submission</strong> is a well-documented Jupyter Notebook. Attach your *.ipynb file under Assignment 2 on Canvas.</li>
</ul>

</div>

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h2 style='color:#2c3e50;'><strong>Section 1: Problem Definition</strong></h2>

<hr style="border-top: 2px solid #3498db; margin-bottom: 20px; margin-top: 20px;">

<h3 style='color:#2c3e50;'>Goal</h3>
<p style='color:#2c3e50;'>Dive into the realm of Molecular Biology and explore how a Language Model (LM) like BERT can be harnessed for protein sequences. </p>

<h3 style='color:#2c3e50;'>Introduction to the Problem Domains</h3>
<ul style='color:#2c3e50;'>
    <li><strong>Proteins</strong> are the workhorses of the body, performing a wide range of functions and biological processes necessary for survival and wellbeing. Understanding their functions can aid in drug discovery, disease diagnosis, and much more.
    Proteins are at least 50, and usually more than 100, amino acids in length and composed of multiple peptide subunits. Each protein consists of a linear sequence of amino acids. The sequence of a protein is usually notated as a string of letters including 22 aminoacids and are varied in length from very short to very long! For instance, in NLP we use words and sentences as sequences, similarly in protein modeling we also work with sequences which looks something like this:
    <br>​
    V​​L​​S​​P​​A​​D​​K​​T​​N​​VK​​A​​A​​W​​G​​K​​V​​G​​A​​H​A​​G​​E​​Y​​G​​A​​E​​A​​L​​E​R​​M​​F​​L​​S​​F​​P​​T​​T​​KT​​Y​​F​​P​​H​​F​​D​​L​​S​​HG​​S​​A​​Q​​V​​K​​G​​H​​G​​K​K​​V​​A​​D​​A​​L​​T​​N​​A​​VA​​H​​V​​D​​D​​M​​P​​N​​A​​L​S​​A​​L​​S​​D​​L​​H​​A​​H​​K​​L​​R​​V​​D​​P​​V​​N​​F​​K​​L​L​​S​​H​​C​​L​​L​​V​​T​​L​​AA​​H​​L​​P​​A​​E​​F​​T​​P​​A​V​​H​​A​​S​​L​​D​​K​​F​​L​​A​​S​​V​​S​​T​​V​​L​​T​​S​​K​​Y​
    </br>
    <img alt="Protein Structure" src="protein.jpeg" width="600" height="400" align="center"/>
    <br>
    This sequence belongs to a structure that looks like this: </br>
    </li>
    <img alt="Protein Structure" src="structure.jpg" width="600" height="400" align="center"/>
    <br>
    You can find more information about proteins <a href="https://www.khanacademy.org/science/biology/macromolecules/proteins-and-amino-acids/v/introduction-to-amino-acids">here</a>.
    </br>
    <li><strong>Signal Peptides (SP)</strong> are short sequences of amino acids, typically between 2 and 50 amino acids in length, within a protein that direct the protein to specific locations within or outside the cell. They act like postal addresses, guiding the cellular delivery machinery to transport the protein to its intended location.Once the protein reaches its destination, the signal peptide is often cleaved off by specific enzymes. In the following image, you can see the difference between a peptide and a protein:
    <br>
    <img alt="protein vs peptide" src="peptide.jpeg" width="600" height="400" align="center"/>
    <br>
    You can find more information about peptides <a href="https://www.cellgs.com/blog/the-difference-between-peptides-and-proteins.html">here</a>.
    <br>
    We have a dataset for signal peptides detection in "data/signal_peptide" directory. In this dataset, if a protein sequence has a signal peptide, it is labeled as 1, otherwise it is labeled as 0.
    </br>
    </li>
    <li><strong>SCOP</strong>, which stands for Structural Classification of Proteins is a largely manual classification of protein structural domains based on similarities of their structures and amino acid sequences. Predicting the structural class of a protein sequence is another fine-tuning task that we will do in this assignment. For each protein sequence, we have its structural class as its label in "data/scop" directory. Labels are categorical and there are 7 classes in total.
    </li>
</ul>
</div>

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Import Libraries and Read Data</strong></h3>


In [ ]:
# Please do not change this cell. Run it without changes.
# This cell imports the required libraries and sets some parameters.
# If you want to change the parameters, you can do so here and mention why you changed them with a comment.

import numpy as np
import pandas as pd
import tensorflow as tf
import os
import math
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

max_length = 256 # size of the input sequences - if a sequence is longer than this, it will be truncated. If it's shorter, it will be padded.
batch_size = 50  # number of sequences that will be given to the model at once
epochs = 10     # number of times the model will see the whole training set
lr = 1e-2      # learning rate
vocab_size = 26  # size of the vocabulary - do not change
embed_size = 128 # size of the embeddings - do not change
n_heads = 4     # number of attention heads - do not change
n_layers = 2   # number of transformer layers - do not change

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# please don't change this cell. Run it without changes.
# this cell is for loading the data needed for the assignment.
# pretrain data is downloaded from UniprotKB-SwissProt (https://www.uniprot.org/downloads) latest release.

pretrain_data = pd.read_csv('/content/drive/MyDrive/data/pretrain.tsv', sep='\t')
# scop data
scop_train_data = pd.read_csv('/content/drive/MyDrive/data/scop/scop.train.csv')
scop_test_data = pd.read_csv('/content/drive/MyDrive/data/scop/scop.test.csv')
# signal peptide data
signalp_train_data = pd.read_csv('/content/drive/MyDrive/data/signal_peptide/signalP_binary.train.csv')
signalp_test_data = pd.read_csv('/content/drive/MyDrive/data/signal_peptide/signalP_binary.test.csv')

In [ ]:
# please don't change this cell. Run it without changes.
# this cell is for looking at the pretrain data to get an idea of what it looks like.

pretrain_data.head()

id                                                seq
0  Q65W17  MKPLVIKLGGVLLDTPAAMENLFTALADYQQNFARPLLIVHGGGCL...
1  Q12ZI9  MFTILTGSQFGDEGKGKIVDLLSKDYDLVVRFQGGDNAGHTVVVGD...
2  O08400  MTRISRSAYAEIYGPTVVGGVGDRVRLADTLLLAEVEKDHTIFGEE...
3  O14232  MFGGELDDAFGVFEGKVPKSLKEESKNSQNSQNSQKIKRTLTDKNA...
4  Q2FQ95  MNILIVNRYGDPDVEEFSYELEKLLHHHGHHTSIYKENLLGEAPPL...

In [ ]:
# please don't change this cell. Run it without changes.
# this cell is for looking at the scop data to get an idea of what it looks like.

scop_train_data.head()

seq label
0  MSPFTGSAAPTPEWRHLRVEITDGVATVTLARPDKLNALTFEAYAD...     c
1  MVVTKLAPDFKAPAVLGNNEVDEHFELSKNLGKNGVILFFWPKDFT...     c
2  MKVGIDAGGTLIKIVQEQDNQRTFKTELTKNIDQVVEWLNQQQIEK...     c
3  LYKLLILDIDGTLRDEVYGIPESAKHAIRLCQKNHCSVVICTGRSM...     c
4  NTSNITFIGGGNMARNIVVGLIANGYDPNRICVTNRSLDKLDFFKE...     c

In [ ]:
# please don't change this cell. Run it without changes.
# this cell is for looking at the signal peptide data to get an idea of what it looks like.

signalp_train_data.head()

label                                                seq
0      0  MLGMIRNSLFGSVETWPWQVLSTGGKEDVSYEERACEGGKFATVEV...
1      1  MQPAKNLLFSSLLFSSLLFSSAARAASEDGGRGPYVQADLAYAAER...
2      0  MDKGEGLRLAATLRQWTRLYGGCHLLLGAVVCSLLAACSSSPPGGV...
3      0  MKFIDEAKIEVAAGKGGNGATSFRREKFVPRGGPDGGDGGKGGSVW...
4      0  MVAGMLMPRDQLRAIYEVLFREGVMVAKKDRRPRSLHPHVPGVTNL...

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Section 1. Tokenize Protein Sequences</strong></h3>

<p style='color:#2c3e50;'>Firstly, you need to have your protein sequences in a format that can be fed into the model. Each amino acid can be represented by a unique token (or ID), similar to how each word is represented by a unique token in NLP tasks. A common representation is to use the single-letter codes for amino acids.</p>
<p style='color:#2c3e50;'>We have <b>22 amino acids: "ACDEFGHIKLMNPQRSTUVWXY" </b>. There also can be invalid amino acids in protein sequences that we can encode them by <b>'OTHER' </b>token. We need tokens <b>'START'</b>, <b>'END'</b>, and <b>'PAD'</b> for showing starting of the protein sequence, end of protein sequence and padding the protein sequence to a fixed size. <p><b> Do not use predefined tokenizer in keras or tensorflow. You need to implement your own tokenizer for protein sequences.</b></p>
</p>
<p>
    <b>Example (padding to size 32):</b> <br>
    <b>Label encoding:</b> {A: 0, C: 1, D: 2, E: 3, F: 4, G: 5, H: 6, I: 7, K: 8, L: 9, M: 10, N: 11, P: 12, Q: 13, R: 14, S: 15, T: 16, V: 17, W: 18, X: 19, Y: 20, OTHER: 21, START: 22, END: 23, PAD: 24} <br>
    <b>Input:</b> "ACDEFGHIKLMNPQRSTUVWXY" <br>
    <b>Output:</b> [22, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10,11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 23, 24, 24, 24, 24, 24, 24, 24, 24, 24] <br>
</p>
</div>


In [ ]:
# fill this function to tokenize a sequence into a list of integers and return the list of integers representing the sequence like the example above.
def tokenize_seq(seq, max_length=256):
    '''
    Tokenize a sequence into a list of integers.
    assign each amino acid to a unique integer and add <START>, <END>, <PAD>, <OTHER> tokens to the vocabulary.
    <START> token is added to the beginning of each sequence
    <END> token is added to the end of each sequence
    <PAD> token is used to pad sequences to the same length (max_length)
    <OTHER> token is used to represent all amino acids that are not in the vocabulary

    Args:
        seq (str): protein sequence
        max_length (int): maximum length of the sequence
    Returns:
        tokenized_seq: list of integers representing the sequence
    '''
    # YOUR CODE HERE

    amino_acids = "ACDEFGHIKLMNPQRSTUVWXY"
    vocab = {aa: i for i, aa in enumerate(amino_acids)}
    vocab['OTHER'] = len(amino_acids)
    vocab['START'] = len(amino_acids) + 1
    vocab['END'] = len(amino_acids) + 2
    vocab['PAD'] = len(amino_acids) + 3

    tokenized_seq = [vocab.get('START')]
    for aa in seq:
        if aa in vocab:
            tokenized_seq.append(vocab[aa])
        else:
            tokenized_seq.append(vocab['OTHER'])


    tokenized_seq.append(vocab['END'])

    # Padding the sequence to the maximum length
    if len(tokenized_seq) < max_length:
        tokenized_seq.extend([vocab['PAD']] * (max_length - len(tokenized_seq)))
    else:
        tokenized_seq = tokenized_seq[:max_length]

    return tokenized_seq



In [ ]:
# please don't change this cell and don't add any print statements to this cell. Run it without changes.
# this cell is tokenizing the pretrain data and converting it to numpy array.

pretrain_seq = [tokenize_seq(seq) for seq in pretrain_data['seq']]
pretrain_seq = np.array(pretrain_seq)
pretrain_seq = pretrain_seq.astype(np.int32)

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h2 style='color:#2c3e50;'><strong>Section 2: Pretrain a Language Model</strong></h2>

<hr style="border-top: 2px solid #3498db; margin-bottom: 20px; margin-top: 20px;">

<ul style='color:#2c3e50;'>
    <li>Implement the attention mechanism</li>
    <li>Implement BERT architecture</li>
    <li>Train the model by the protein sequences.</li>
</ul>

</div>

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Task 2.1: Implementing the Self-Attention Class</strong></h3>

<p style='color:#2c3e50;'>Implement a multihead self-attention mechanism that will be the building block for BERT model.</p>

<p style='color:#2c3e50;'><strong>Multihead self-attention</strong> is a mechanism at the heart of Transformer models, which are widely used for various natural language processing tasks. It allows the model to focus on different positions of the input sequence simultaneously when producing an output. Here's a simplified explanation:
<p style='color:#2c3e50;'>
<strong>Self-Attention:</strong> This component allows each position in a sequence to attend to all positions within the same sequence when computing the representation of itself. It helps the model capture context from the entire sequence.
<p style='color:#2c3e50;'>
<strong>Multihead:</strong> The 'multihead' part means that the self-attention process is duplicated multiple times (each 'head' being one instance). Each head attends to information from different representational spaces at different positions. This means that instead of having a single set of attention weights for each position, you have multiple sets, allowing the model to capture a diverse range of information which enhances its learning capacity for complex patterns in data.
<p style='color:#2c3e50;'>
<strong>Mechanism Overview:</strong> In practice, for each head, the input sequence is linearly transformed to a set of queries, keys, and values using learned weights. Then, the attention scores are computed by a <b>scaled dot-product of queries and keys</b>, which are used to weight the values. The outputs of each head are concatenated and once again linearly transformed to produce the final output. </p>
<img alt="Multihead Self-Attention" src="attention.jpg" width="600" height="400" align="center"/>
<p> You can read more about attention mechanism <a href="https://towardsdatascience.com/transformers-explained-visually-part-3-multi-head-attention-deep-dive-1c1ff1024853">here</a>.</p>
<p>
An attention mask in the context of self-attention mechanisms is used to prevent certain positions from being attended to. For instance, when processing sequences of different lengths, padding tokens are added to achieve uniformity in length. An attention mask can be applied to ensure that these padding tokens do not influence the model's output. It's also used to enforce causality in sequence-to-sequence tasks, ensuring that predictions for a certain position can't depend on future positions. Although it is necessary to use attention masks in some cases, we will not use them in this assignment.
</div>

In [ ]:
import tensorflow as tf

class MultiSelfAttentionHead(tf.keras.layers.Layer):
    def __init__(self, num_heads=4, key_dim=64, **kwargs):
        super(MultiSelfAttentionHead, self).__init__(**kwargs)
        self.num_heads = num_heads
        self.key_dim = key_dim
        self.scale = key_dim ** -0.5
        self.depth = key_dim // num_heads

    def split_heads(self, x, batch_size):
        # split the last dimension into (num_heads, depth) and transpose the result such
        # that the shape is (batch_size, num_heads, seq_len, depth)
        # Fill variable x
        input_shape=x.shape
        #x = tf.reshape(x, (batch_size,self.num_heads,input_shape[1], self.depth))
        x=tf.reshape(x, (tf.shape(x)[0], -1, self.num_heads, self.depth))
        x = tf.transpose(x, perm=[0, 2, 1, 3])


        return x


    def build(self, inputs_s):
        # Create trainable weight matrices for Q, K, V, and the final output

        #self.wq = tf.Variable(tf.random.normal((inputs_s[-1], self.key_dim)), dtype=tf.float32)
        #print(self.wq.shape)
        self.wq= tf.keras.layers.Dense(units=self.key_dim , use_bias=True)
        self.wk = tf.keras.layers.Dense(units=self.key_dim , use_bias=True)
        self.wv= tf.keras.layers.Dense(units=self.key_dim , use_bias=True)
        self.wo=tf.keras.layers.Dense(units=inputs_s[-1], use_bias=True)
        assert self.key_dim % self.num_heads == 0, "key_dim must be divisible by num_heads"
        super(MultiSelfAttentionHead, self).build(inputs_s)

    def call(self, inputs):
        batch_size = tf.shape(inputs)[0]
        # Linear projections of Q, K, V (Gray Linear layers in the begining of the right image above)
        # Fill variables query, key, value
        #q = tf.matmul(inputs, self.wq)  # (batch_size, seq_len, key_dim)

        q=self.wq(inputs)
        k = self.wk(inputs)  # (batch_size, seq_len, key_dim)
        v = self.wv(inputs) # (batch_size, seq_len, key_dim)

        # Split the heads using split_heads function
        # Fill variables query, key, value


        query = self.split_heads(q, batch_size) # (batch_size, num_heads, seq_len_q, depth)
        key = self.split_heads(k, batch_size) # (batch_size, num_heads, seq_len_k, depth)
        value = self.split_heads(v, batch_size) # (batch_size, num_heads, seq_len_v, depth)

        # Linear projection between Q and K (in the left image above)
        # Fill variable linear_qk

        linear_qk = tf.matmul(query, key, transpose_b=True)


        # Scale Linear projection between Q and K using self.scale (in the left image above)
        # Fill variable scaled_linear_qk

        scaled_linear_qk = linear_qk * self.scale


        # Apply softmax to the last axis of scaled_linear_qk
        # Fill variable attention_weights

        attention_weights = tf.nn.softmax(scaled_linear_qk, axis=-1)


        # Matmul of attention_weights and value (in the left image above)
        # Fill variable context

        context = tf.matmul(attention_weights, value)# (batch_size, num_heads, seq_len_q, depth)


        # 'Concatenate' heads (in the right image above) using tf.transpose and tf.reshape
        # Fill variable context


        context = tf.transpose(context, perm=[0, 2, 1, 3])

        context = tf.reshape(context, (batch_size, -1, self.key_dim))


        # Final linear projection (Gray Linear layer in the end of the right image above) between context and self.wo
        # Fill variable outputs

        outputs = self.wo(context)


        return outputs

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Task 2.2: Implementing the BERT Architecture</strong></h3>

<p style='color:#2c3e50;'>Implement BERT architecture.</p>
<p> You can read more about BERT in the original <a href="https://arxiv.org/pdf/1810.04805.pdf">paper</a>.</p>
<p>
BERT is based on encoder transformer. Encoder consists of multi-head attention and feed-forward neural network. In original Transformers paper <a href="https://arxiv.org/pdf/1706.03762.pdf">Attention Is All You Need</a> and in this assignment, the stack of identical encoder layers is used.
</p>
<img alt="Encoder transformer" src="encoder.png" width="300" height="400" align="center"/>
<p> In BERT function, you need to implement the following steps:</p>
<p>
<strong>Inputs:</strong> The function takes several parameters like vocabulary size, embedding size, number of layers, attention heads, and maximum sequence length. Additionally, it expects an input sequence and an attention mask when called.
<p>
<strong>Embedding Layer:</strong> The model begins by mapping the input sequence of tokens to vectors using an embedding layer.
<p>
<strong>Encoder Layers:</strong> It then passes the embedded input through a series of identical layers. Each layer consists of:
1. A multi-head self-attention mechanism that allows the model to capture different aspects of the data. The attention output is combined with the original embeddings using a residual connection followed by layer normalization.
2. A feed-forward neural network (FFN) with two dense layers is applied after each attention mechanism, also followed by a residual connection and layer normalization.
<p>
<strong>Outputs:</strong> After processing the input through the specified number of encoder layers, the function applies a dense layer with softmax activation to produce a probability distribution over the vocabulary for each position in the input sequence.
<p>
<strong>Model Creation:</strong> Finally, it encapsulates the complete architecture within a Keras Model object, which can be compiled and trained.

</div>

In [ ]:

import tensorflow as tf
from tensorflow.keras.layers import Input, Embedding, MultiHeadAttention, Dense
from tensorflow.keras.layers import LayerNormalization, Dropout, Flatten

def BERT(vocab_size, embed_size=128, n_layers=2, n_heads=4, max_length=256):
    '''
    BERT model.

    Args:
        vocab_size (int): vocabulary size (number of tokens)
        embed_size (int): embedding size (dimension of the token embedding)
        n_layers (int): number of layers in the encoder stack
        n_heads (int): number of attention heads in each multi-head attention layer
        max_length (int): maximum length of the sequence

    Inputs:
        input_seq (tf.Tensor): input tensor with shape [Batch_size, max_length];
            No positional encoding is added to the input sequence

    Outputs:
        output (tf.Tensor): output tensor with shape [Batch_size, max_length, vocab_size]
        return model (tf.keras.Model): BERT model

    Model Architecture (as described above):
        1. Input Layer
        2. Embedding Layer
        3. Multi-Head Self-Attention Layers
        4. Add and Norm Layer
        5. Feed-Forward Layers
        6. Add and Norm Layer
        7. Dense Layer (vocab_size) with softmax activation
        Steps 3-6 are repeated n_layers times.
    '''

    # Input layer
    #inputs = Input(shape=(max_length,), dtype=tf.int32, name='input_seq')

    # Embedding layer
    input= tf.keras.Input(shape=(max_length,), batch_size=50, name='input_seq')

    #batch =Input(shape=(max_length,), dtype=tf.float32, name='input_seq')
    embedding = Embedding(input_dim=vocab_size, output_dim=embed_size)(input)




    # Multi-Head Self-Attention Layers (Repeat n_layers times)
    for _ in range(n_layers):

        attention_layer = MultiSelfAttentionHead(num_heads=4, key_dim=64)
        attention = attention_layer(embedding)

        add_norm_1 = LayerNormalization(epsilon=1e-6)(embedding + attention)

        feed_forward1 = Dense(units=embed_size, activation='relu')(add_norm_1)
        feed_forward2 = Dense(units=embed_size, use_bias=True)(feed_forward1)
        add_norm_2 = LayerNormalization(epsilon=1e-6)(add_norm_1 + feed_forward2)
        embedding = add_norm_2


    # Output layer

    output = Dense(units=vocab_size, activation='softmax')(embedding)

    # Build model
    model = tf.keras.Model(inputs=input, outputs=output)
    return model


<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">
<h3 style='color:#2c3e50;'><strong>Section 3: Implementing Noised Tokens Prediction (NTP)</strong></h3>

<p style='color:#2c3e50;'>In previous cell, you implemented BERT architecture. In main version of BERT, we use Masked Language Models (MLMs) and Next Sentence Prediction (NSP) as pretraining tasks.
<p>
In Masked Language Modeling, random tokens in a sentence are replaced with a special token (e.g., "[MASK]"). The model's task is to predict the original token based on the context provided by the other non-masked tokens in the input. This allows the model to develop a deep understanding of language structure and word relationships.
<p>
The Next Sentence Prediction task involves presenting the model with two sentences and requiring it to predict whether the second sentence follows the first in the original text. This helps the model to learn relationships between consecutive sentences, which is important for understanding the coherence and flow of paragraphs.
<br>
You can read more about BERT pretraining tasks in the original <a href="https://arxiv.org/pdf/1810.04805.pdf">paper</a> or <a href="https://towardsdatascience.com/bert-explained-state-of-the-art-language-model-for-nlp-f8b21a9b6270">here</a>.
<p>
In this assignment, however, we want to define a new task called <b>Noised Tokens Prediction (NTP)</b> and use it instead of MLMs and NSP tasks. In this task, we randomlely change tokens in the input sequence to get a noisy sequence to get it to the model and want the model to predict the original sequence. For example, if the input sequence is "ABCDEF", with some probability, it is changed into "ABXDEF"; so the input of the model would be "ABXDEF" and the output should be "ABCDEF". You need to implement this task in the following cell. Feel free to use numpy library for randomization.</p>
</p>

</div>

In [ ]:
import numpy as np

def noise_tokens(encoded_seqs, p_token_noise=0.15, n_tokens=vocab_size):
    '''
    Add noise to the input sequences.
    Args:
        encoded_seqs (np.array): encoded sequences with shape [Batch_size, max_length]
        p_token_noise (float): probability of adding noise to each token
        n_tokens (int): number of tokens in the vocabulary
    Returns:
        noisy_encoded_seqs (np.array): encoded sequences with noise with shape [Batch_size, max_length]
        original_seqs (np.array): original encoded sequences with shape [Batch_size, max_length]
    '''
    # Copy the original sequences to avoid modifying the input
    noisy_encoded_seqs = np.copy(encoded_seqs)

    # Create a mask with True values based on the probability
    mask = np.random.rand(*encoded_seqs.shape) < p_token_noise

    # Generate random noisy tokens for the masked positions
    noisy_tokens = np.random.randint(0, n_tokens, size=mask.sum())

    # Apply the noise to the corresponding positions in the sequences
    noisy_encoded_seqs[mask] = noisy_tokens

    return noisy_encoded_seqs, encoded_seqs

# Example usage:
# Replace 'encoded_seqs' with your actual encoded sequences
# Specify the appropriate values for 'p_token_noise' and 'n_tokens'
# noisy_seqs, original_seqs = noise_tokens(encoded_seqs, p_token_noise=0.15, n_tokens=vocab_size)


<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Section 4: Pretraining the Model</strong></h3>

<p style='color:#2c3e50;'>Create an instance of your BERT model and train it on pretrain data.</p>

</div>

In [ ]:
import tensorflow as tf

def pretraining(model, noisy_pretrain_seq, pretrain_seq, epochs, batch_size, lr=1e-2):
    '''
    Pretrain the model: Input is noised sequences and output is original sequences.
    Args:
        model (tf.keras.Model): BERT model
        noisy_pretrain_seq (np.array): encoded sequences with noise with shape [Batch_size, max_length]
        pretrain_seq (np.array): original encoded sequences with shape [Batch_size, max_length]
        epochs (int): number of epochs
        batch_size (int): batch size
        lr (float): learning rate
    Returns:
        model (tf.keras.Model): pretrained BERT model
    '''

    # Define the loss function and optimizer
    loss_fn = tf.keras.losses.CategoricalCrossentropy()
    optimizer = tf.keras.optimizers.Adam(learning_rate=lr)
    model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=lr),
                         loss=loss_fn,
                         metrics=['accuracy'])


    # Compile the model
    #model.compile(optimizer=optimizer, loss=loss_fn)

    original_sequences_one_hot = tf.keras.utils.to_categorical(pretrain_seq, num_classes=vocab_size)


    # Train the model


    model.fit(noisy_pretrain_seq, original_sequences_one_hot, epochs=epochs, batch_size=50)

    return model

# Example usage:
# Replace 'YourBERTModel' with the actual class or instance of your BERT model
# Replace 'noisy_pretrain_seq' and 'pretrain_seq' with your actual noisy and original sequences
# Specify the appropriate values for 'epochs', 'batch_size', and 'lr'
# model = pretraining(YourBERTModel, noisy_pretrain_seq, pretrain_seq, epochs=10, batch_size=32, lr=1e-3)


In [ ]:
# please don't change this cell and don't add any print statements to this cell. Run it without changes.
# this cell is for training the model on the pretrain data.

model = BERT(vocab_size=vocab_size, embed_size=128, n_layers=2, n_heads=4, max_length=256)
model.summary()

noisy_pretrain_seq, pretrain_seq = noise_tokens(pretrain_seq)

pretrained_model = pretraining(model, noisy_pretrain_seq, pretrain_seq, epochs=1, batch_size=50)

Model: "model"
__________________________________________________________________________________________________
 Layer (type)                Output Shape                 Param #   Connected to                  
 input_seq (InputLayer)      [(50, 256)]                  0         []                            
                                                                                                  
 embedding (Embedding)       (50, 256, 128)               3328      ['input_seq[0][0]']           
                                                                                                  
 multi_self_attention_head   (50, 256, 128)               33088     ['embedding[0][0]']           
 (MultiSelfAttentionHead)                                                                         
                                                                                                  
 tf.__operators__.add (TFOp  (50, 256, 128)               0         ['embedding[0][0]',       

In [ ]:
pretrained_model.save_weights("model_pretrained.weights.h5")

In [ ]:
# please don't change this cell and don't add any print statements to this cell. Run it without changes.
# this cell is for saving the pretrained model.

if not os.path.exists('model'):
    os.makedirs('model')
pretrained_model.save_weights('model/model_pretrained.h5')

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h2 style='color:#2c3e50;'><strong>Section 5: Fine-tuning Large Language Models</strong></h2>

<hr style="border-top: 2px solid #3498db; margin-bottom: 20px; margin-top: 20px;">

<ul style='color:#2c3e50;'>
    <li>Prepare the data for fine-tuning: Tokenize protein sequences in both train and test data of signal peptides detection and structure classification tasks. Then, convert them to numpy arrays. (like what you did in section 1 for pretraining data)</li>
    <li>Fine-tune the model on tasks signal peptides detection (binary classification) and structure classification (multi-class classification). By fine-tuning, we mean the process of adjusting the parameters of the pre-trained model to suit a specific task or dataset. Through this process, we can enhance the model's ability to perform a specialized task more accurately.</li>
    <li>For each task, you need to call the pretrained model and add a classification layer on top of it. Then, you need to freeze the pretrained model and train the classification layer for 10 epochs. After that, you need to unfreeze the pretrained model and train the whole model for 5 epochs.</li>
    </li>
</ul>

</div>

In [ ]:
# please don't change this cell and don't add any print statements to this cell. Run it without changes.
# this cell is for connecting the layers of the pretrained model to the layers of the finetuned model.
# the output of the last layer of the pretrained model is considered as the input of the finetuned model.

def concat_layers_for_finetuning(model):
    # Create a list to hold the layers to be concatenated

    return tf.keras.Model(inputs=model.inputs, outputs=model.layers[-1].output)

In [ ]:
# please don't change this cell
# this cell is for loading the pretrained model and keep the output of the last layer to be used for finetuning.

model_SP = BERT(vocab_size=vocab_size, embed_size=embed_size, n_layers=n_layers, n_heads=n_heads, max_length=max_length)
model_SP.load_weights(("model_pretrained.weights.h5"))
concat_model_SP = concat_layers_for_finetuning(model_SP)


In [ ]:
# Fill this function to fine-tune the concat_model_SP on peptide detection task (binary classification)

'''
Fine-tune the model on peptide detection task.
Parameters:
    epochs (int): number of epochs
    batch_size (int): batch size
    lr (float): learning rate
    n_classes (int): number of classes

What you need to do in this cell:
    1. Tokenize the sequences in scop_train_data
    2. Make the labels in scop_train_data as a numpy array
    3. Make the layers of the concat_model_SP non-trainable
    4. Add a dense layer with a proper activation to the end of the concat_model_SP
    5. Define a new model with the input of concat_model_SP and the output of the dense layer
    6. Define the loss function and optimizer
    7. Compile the model with the loss function and optimizer
    8. Fit the model on the training data for 10 epochs
    9. Make all layers of the model trainable
    10. Fit the model on the training data for 5 epochs
    11. Save the new model in the model directory with the name model_finetuned_SP.h5
    '''

# YOUR CODE HERE
import os
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Dense

from tensorflow.keras.optimizers import Adam
from tensorflow.keras.losses import BinaryCrossentropy
from tensorflow.keras.losses import CategoricalCrossentropy
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

# Assuming 'model_directory' already exists


def fine_tune_peptide_detection(concat_model_SP, signalp_train_data, epochs=15, batch_size=50, lr=0.001, n_classes=2):
    # Tokenize sequences


    signalp_train_seq = [tokenize_seq(seq) for seq in  signalp_train_data['seq']]
    signalp_train_seq  = np.array(signalp_train_seq )
    X_train= signalp_train_seq .astype(np.int32)

    y_train = (signalp_train_data['label'])
    y_train = tf.keras.utils.to_categorical( y_train, num_classes=n_classes)


    # Make layers of concat_model_SP non-trainable
    for layer in concat_model_SP.layers:
        layer.trainable = False

    output_layer = Dense(n_classes, activation='sigmoid')(concat_model_SP.layers[-1].output[:,0])


    # Define a new model

    model = Model(inputs=concat_model_SP.input, outputs= output_layer)

    # Define loss function and optimizer
    loss_function =BinaryCrossentropy()
    optimizer = Adam(learning_rate=lr)

    # Compile the model
    model.compile(optimizer=optimizer, loss=loss_function, metrics=['accuracy'])

    # Fit the model on the training data for 10 epochs

    model.fit(X_train, y_train, epochs=10, batch_size=batch_size)
    # Make all layers of the model trainable
    for layer in model.layers:
        layer.trainable = True

    # Fit the model on the training data for additional 5 epochs

    model.fit(X_train, y_train, epochs=5, batch_size=batch_size)

    # Save the entire model
    model.save_weights("model_finetuned_SP.weights.h5")

    return model


# Call the function with the provided parameters
Finetune_SP=fine_tune_peptide_detection(concat_model_SP, signalp_train_data, epochs=15, batch_size=50, lr=0.001, n_classes=2)



Epoch 1/10
333/333 [==============================] - 101s 293ms/step - loss: 0.5484 - accuracy: 0.8375
Epoch 2/10
333/333 [==============================] - 100s 299ms/step - loss: 0.4752 - accuracy: 0.8375
Epoch 3/10
333/333 [==============================] - 98s 293ms/step - loss: 0.4519 - accuracy: 0.8375
Epoch 4/10
333/333 [==============================] - 99s 297ms/step - loss: 0.4455 - accuracy: 0.8375
Epoch 5/10
333/333 [==============================] - 100s 299ms/step - loss: 0.4441 - accuracy: 0.8375
Epoch 6/10
333/333 [==============================] - 100s 299ms/step - loss: 0.4439 - accuracy: 0.8375
Epoch 7/10
333/333 [==============================] - 99s 299ms/step - loss: 0.4438 - accuracy: 0.8375
Epoch 8/10
333/333 [==============================] - 97s 292ms/step - loss: 0.4438 - accuracy: 0.8375
Epoch 9/10
333/333 [==============================] - 97s 291ms/step - loss: 0.4437 - accuracy: 0.8375
Epoch 10/10
333/333 [==============================] - 97s 292ms/step

In [ ]:
# please don't change this cell
# this cell is for loading the pretrained model and keep the output of the last layer to be used for finetuning.

model_SC = BERT(vocab_size=vocab_size, embed_size=128, n_layers=2, n_heads=4, max_length=max_length)
model_SC.load_weights(("model_pretrained.weights.h5"))
model_concat_SC = concat_layers_for_finetuning(model_SC)

In [ ]:
# Fill this function to fine-tune the model_concat_SC on structure classification task (multi-class classification)
'''
Fine-tune the model on peptide detection task.
Parameters:
    epochs (int): number of epochs
    batch_size (int): batch size
    lr (float): learning rate
    n_classes (int): number of classes

What you need to do in this cell:
    1. Tokenize the sequences in scop_train_data
    2. Make the labels in scop_train_data as a numpy array
    3. Make the layers of the model_concat_SC non-trainable
    4. Add a dense layer with a proper activation to the end of the model_concat_SC
    5. Define a new model with the input of model_concat_SC and the output of the dense layer
    6. Define the loss function and optimizer
    7. Compile the model with the loss function and optimizer
    8. Fit the model on the training data for 10 epochs
    9. Make all layers of the model trainable
    10. Fit the model on the training data for 5 epochs
    11. Save the new model in the model directory with the name model_finetuned_SC.h5
'''

# YOUR CODE HERE
# Fill this function to fine-tune the concat_model_SP on peptide detection task (binary classification)

# YOUR CODE HERE
import os
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Dense

from tensorflow.keras.optimizers import Adam
from tensorflow.keras.losses import BinaryCrossentropy
from tensorflow.keras.losses import CategoricalCrossentropy
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

# Assuming 'model_directory' already exists


def fine_tune_structural_classification(concat_model_SC, scop_train_data, epochs=15, batch_size=50, lr=0.001, n_classes=2):
    # Tokenize sequences


    scop_train_seq = [tokenize_seq(seq) for seq in  scop_train_data['seq']]
    scop_train_seq = np.array(scop_train_seq)
    X_train= scop_train_seq.astype(np.int32)

    y_train = (scop_train_data['label'])
    label_mapping = {'a': 0, 'b': 1, 'c': 2,'d': 3, 'e': 4, 'f': 5}
    y_train = y_train.map(label_mapping)
    y_train = y_train.fillna(0)
    y_train = to_categorical(y_train, num_classes=6)


    # Make layers of concat_model_SC non-trainable
    for layer in concat_model_SC.layers:
        layer.trainable = False

    # Add a dense layer to the end of concat_model_SP
    output_layer = Dense(n_classes, activation='sigmoid')(concat_model_SC.layers[-1].output[:,0])
    # Define a new model
    model = Model(inputs=concat_model_SC.input, outputs= output_layer)

    #print(output_layer.shape,"opt")
    # Define loss function and optimizer
    loss_function =CategoricalCrossentropy()
    optimizer = Adam(learning_rate=lr)

    # Compile the model
    model.compile(optimizer=optimizer, loss=loss_function, metrics=['accuracy'])

    # Fit the model on the training data for 10 epochs

    model.fit(X_train, y_train, epochs=10, batch_size=batch_size)
    # Make all layers of the model trainable
    for layer in model.layers:
        layer.trainable = True

    # Fit the model on the training data for additional 5 epochs

    model.fit(X_train, y_train, epochs=5, batch_size=batch_size)

    # Save the entire model
    model.save_weights("model_finetuned_SC.weights.h5")

    model.summary()
    return model

# Call the function with the provided parameters
Model_finetune_SC=fine_tune_structural_classification(model_concat_SC , scop_train_data, epochs=15, batch_size=50, lr=0.001, n_classes=6)




Epoch 1/10
314/314 [==============================] - 99s 308ms/step - loss: 1.6809 - accuracy: 0.2807
Epoch 2/10
314/314 [==============================] - 95s 301ms/step - loss: 1.5427 - accuracy: 0.3807
Epoch 3/10
314/314 [==============================] - 93s 295ms/step - loss: 1.4836 - accuracy: 0.3803
Epoch 4/10
314/314 [==============================] - 93s 298ms/step - loss: 1.4525 - accuracy: 0.3964
Epoch 5/10
314/314 [==============================] - 111s 353ms/step - loss: 1.4325 - accuracy: 0.4122
Epoch 6/10
314/314 [==============================] - 93s 297ms/step - loss: 1.4178 - accuracy: 0.4152
Epoch 7/10
314/314 [==============================] - 132s 420ms/step - loss: 1.4060 - accuracy: 0.4216
Epoch 8/10
314/314 [==============================] - 100s 319ms/step - loss: 1.3961 - accuracy: 0.4254
Epoch 9/10
314/314 [==============================] - 130s 415ms/step - loss: 1.3873 - accuracy: 0.4268
Epoch 10/10
314/314 [==============================] - 113s 360ms/ste

In [ ]:
Model_finetune_SC.save_weights("model_finetune_SC.weights.h5")

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h2 style='color:#2c3e50;'><strong>Section 6: Prediction and Evaluation</strong></h2>

<hr style="border-top: 2px solid #3498db; margin-bottom: 20px; margin-top: 20px;">

<ul style='color:#2c3e50;'>
    <li>After finetuning, do <b>prediction</b> task on test data. Then, report <b>f1 score, precision, recall, and accuracy </b> for both datasets.
    You can read more about these metrics <a href="https://scikit-learn.org/stable/modules/model_evaluation.html#classification-metrics">here</a>.</li>
</ul>

</div>

<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Task 6.1: Prediction </strong></h3>

<p style='color:#2c3e50;'> Make prediction on both test datasets.</p>

</div>

In [ ]:
# testing on signal peptide data; save the predicted labels, true labels and sequences to a csv file
# fill this cell to create y_pred_SP using the model_finetuned_SP

# YOUR CODE HERE


# Create a model with the same architecture as the one used for saving weights
# Replace this with your actual model architecture
n_classes=2

finetune_SP = BERT(vocab_size=vocab_size, embed_size=embed_size, n_layers=n_layers, n_heads=n_heads, max_length=max_length)

finetune_SP.load_weights(("model_finetuned_SP.weights.h5"))
concat_finetune_SP = concat_layers_for_finetuning(finetune_SP)

output_layer = Dense(n_classes, activation='sigmoid')(concat_finetune_SP.layers[-1].output[:,0])

model = Model(inputs=concat_finetune_SP.input, outputs= output_layer)
signalp_test_seq = [tokenize_seq(seq) for seq in signalp_test_data['seq']]
signalp_test_seq = np.array(signalp_test_seq)
X_test= signalp_test_seq.astype(np.int32)
y_pred=model.predict(X_test,batch_size=50)

y_pred_SP = np.argmax(y_pred, axis=1)

print(len(y_pred_SP))


# save the predicted labels, true labels and sequences to a csv file
signalp_test_data['pred_label'] = y_pred_SP

result_df = pd.DataFrame({
    'seq': signalp_test_data['seq'],
    'label': signalp_test_data['label'],  # Replace with your actual column names
    'pred_label': y_pred_SP.tolist()
})
result_df.to_csv('signalp_pred.csv', index=False)

#signalp_test_data.to_csv('signalp_pred.csv', index=False)

84/84 [==============================] - 33s 370ms/step
4152


In [ ]:
# testing on scop data ; save the predicted labels, true labels and sequences to a csv file
# fill this cell to create y_pred_SC using the model_finetuned_SC
n_classes=6
finetune_SC = BERT(vocab_size=vocab_size, embed_size=embed_size, n_layers=n_layers, n_heads=n_heads, max_length=max_length)

finetune_SC.load_weights(("model_finetuned_SC.weights.h5"))
concat_finetune_SC = concat_layers_for_finetuning(finetune_SC)
output_layer = Dense(n_classes, activation='sigmoid')(concat_finetune_SC.layers[-1].output[:,0])
model = Model(inputs=concat_finetune_SC.input, outputs= output_layer)

scop_test_seq = [tokenize_seq(seq) for seq in scop_test_data['seq']]
scop_test_seq = np.array(scop_test_seq)
X_test= scop_test_seq.astype(np.int32)

y_train = (scop_test_data['label'])
label_mapping = {'a': 0, 'b': 1, 'c': 2,'d': 3, 'e': 4, 'f': 5}
y_train = y_train.map(label_mapping)
y_train = y_train.fillna(0)
y_train = y_train.astype(int)

y_pred=model.predict(X_test,batch_size=50)

y_pred_SC = np.argmax(y_pred, axis=1)




scop_test_data['pred_label'] = y_pred_SC

result_df1 = pd.DataFrame({
    'seq': scop_test_data['seq'],
    'label': y_train.tolist(),  # Replace with your actual column names
    'pred_label': y_pred_SC.tolist()
})
result_df1.to_csv('scop_pred.csv', index=False)


# save the predicted labels, true labels and sequences to a csv file


79/79 [==============================] - 32s 389ms/step


<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Task 6.2: Evaluation</strong></h3>

<p style='color:#2c3e50;'>Calculate F1 score, precision, recall and accuracy for both test datasets.</p>

</div>

In [ ]:
# Fill this cell to evaluate the performance of model_finetuned_SP on signal peptide test data
# You can use sklearn library to calculate these metrics.

# accuracy
# precision
# recall
# f1-score

# YOUR CODE HERE
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Assuming you have the true labels (y_true) and predicted labels (y_pred) for your test data
# Replace y_true and y_pred with your actual data

# Calculate accuracy
import pandas as pd

# Assuming your CSV file is named 'your_file.csv'
#file_path = '/signalp_pred.csv'  # Adjust the file path accordingly

# Read the CSV file into a DataFrame
df = pd.read_csv('signalp_pred.csv')

accuracy = accuracy_score(df.label,df.pred_label)

# Calculate precision
precision = precision_score(df.label,df.pred_label)

# Calculate recall
recall = recall_score(df.label,df.pred_label)

# Calculate F1-score
f1 = f1_score(df.label,df.pred_label)

# Print the metrics
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")





Accuracy: 0.1623
Precision: 0.1623
Recall: 1.0000
F1-score: 0.2793


In [ ]:
# Fill this cell to evaluate the performance of model_finetuned_SP on signal peptide test data
# You can use sklearn library to calculate these metrics.

# accuracy
# precision
# recall
# f1-score

# YOUR CODE HERE
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Assuming you have the true labels (y_true) and predicted labels (y_pred) for your test data
# Replace y_true and y_pred with your actual data

# Calculate accuracy
import pandas as pd

# Assuming your CSV file is named 'your_file.csv'
#file_path = '/signalp_pred.csv'  # Adjust the file path accordingly

# Read the CSV file into a DataFrame
df1 = pd.read_csv('scop_pred.csv')

# Display the first few rows of the DataFrame

accuracy = accuracy_score(df1.label,df1.pred_label)
# Calculate precision
precision = precision_score(df1.label, df1.pred_label, average='weighted')


# Calculate recall
recall = recall_score(df1.label, df1.pred_label, average='weighted')


# Calculate F1-score
f1 = f1_score(df1.label, df1.pred_label, average='weighted')


# Print the metrics
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")




Accuracy: 0.2379
Precision: 0.0566
Recall: 0.2379
F1-score: 0.0915


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


<div style="background-color:#f9f9f9; padding:20px; border:1px solid #ddd; border-radius:8px; font-family:Verdana;">

<h3 style='color:#2c3e50;'><strong>Section 7: Experimentation</strong></h3>

<p style='color:#2c3e50;'>In the realm of natural language processing, one essential aspect is the reliance on vast datasets for training language models. With limited access to resources like GPUs or memory, we cannot pretrain our own model in this assignment on a huge dataset. Instead, we can use available pretrained models for protein sequences to check if a LLM can improve the performance. In this section of our study, we aim to utilize a pre-trained language model named <b>DistilProtBert</b>, which is accessible through the popular <b>Transformers</b> library.
The key objective here is to employ this pre-trained DistilProtBert model and fine-tune it, just as we did in the previous section and compare the performance of the fine-tuned model with and without using pretrained model.
Once you have completed the fine-tuning process, it is important to analyze and record any noticeable differences in the model's performance. This entails a comprehensive evaluation of how the fine-tuned model compares to its performance before fine-tuning. You should pay attention to various metrics such as accuracy, precision, recall, and F1-score, among others, to understand the overall impact of the fine-tuning process.
In this section, you are free to use any libraries or pretrained models. However, the smallest pretrained model available for preotein sequences is DistilProtBert.
</p>
<p>
You can read more about DistilProtBert <a href="https://huggingface.co/yarongef/DistilProtBert">here</a>.
<p> What you need to do in this section is: </p>
<p>
1. Select one of two finetuning tasks from data directory.</p>
<p>
2. Finetune DistilProtBert on the selected dataset and report the performance of the model on test data. You need to report <b>f1 score, precision, recall, and accuracy </b>. </p>
<p>
3. Train your classifier of previous step without using pretrained model (DisilProtBert). Report the performance of the model on test data. You need to report <b>f1 score, precision, recall, and accuracy </b>.
</p>
<p>
4. Compare the performance of the two models and report your findings.
</p>
</div>

In [23]:
%pip install transformers[torch]
%pip install accelerate -U

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 265.7/265.7 kB 5.8 MB/s eta 0:00:00


In [24]:
from transformers import DistilBertModel, DistilBertTokenizer, \
Trainer, TrainingArguments, DistilBertConfig
import torch
from torch.utils.data import Dataset, DataLoader
import torch.nn.functional as F
import pandas as pd
import numpy as np
import torch.nn as nn

In [25]:

# Load the pretrained model and tokenizer from yarongef/DistilProtBert
tokenizer = DistilBertTokenizer.from_pretrained("yarongef/DistilProtBert", do_lower_case=False)
model_distilprotbert = DistilBertModel.from_pretrained("yarongef/DistilProtBert")

tokenizer_config.json:   0%|          | 0.00/86.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/80.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/589 [00:00<?, ?B/s]

The tokenizer class you load from this checkpoint is not the same type as the class this function is called from. It may result in unexpected tokenization. 
The tokenizer class you load from this checkpoint is 'BertTokenizer'. 
The class this function is called from is 'DistilBertTokenizer'.
You are using a model of type bert to instantiate a model of type distilbert. This is not supported for all configurations of models and can yield errors.


pytorch_model.bin:   0%|          | 0.00/924M [00:00<?, ?B/s]

Some weights of DistilBertModel were not initialized from the model checkpoint at yarongef/DistilProtBert and are newly initialized: ['transformer.layer.2.sa_layer_norm.bias', 'transformer.layer.3.ffn.lin1.bias', 'transformer.layer.10.sa_layer_norm.bias', 'transformer.layer.8.ffn.lin2.weight', 'transformer.layer.3.output_layer_norm.bias', 'transformer.layer.2.attention.out_lin.bias', 'transformer.layer.0.attention.q_lin.weight', 'transformer.layer.1.sa_layer_norm.weight', 'transformer.layer.8.attention.v_lin.weight', 'transformer.layer.11.ffn.lin2.bias', 'transformer.layer.8.ffn.lin1.weight', 'transformer.layer.4.attention.q_lin.weight', 'transformer.layer.1.attention.out_lin.weight', 'transformer.layer.13.attention.q_lin.weight', 'transformer.layer.4.ffn.lin2.weight', 'transformer.layer.10.attention.v_lin.bias', 'transformer.layer.0.sa_layer_norm.bias', 'transformer.layer.10.attention.v_lin.weight', 'transformer.layer.1.ffn.lin1.weight', 'transformer.layer.0.ffn.lin2.bias', 'transform

In [ ]:
model_distilprotbert

DistilBertModel(
  (embeddings): Embeddings(
    (word_embeddings): Embedding(30, 1024, padding_idx=0)
    (position_embeddings): Embedding(40000, 1024)
    (LayerNorm): LayerNorm((1024,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (transformer): Transformer(
    (layer): ModuleList(
      (0-14): 15 x TransformerBlock(
        (attention): MultiHeadSelfAttention(
          (dropout): Dropout(p=0.1, inplace=False)
          (q_lin): Linear(in_features=1024, out_features=1024, bias=True)
          (k_lin): Linear(in_features=1024, out_features=1024, bias=True)
          (v_lin): Linear(in_features=1024, out_features=1024, bias=True)
          (out_lin): Linear(in_features=1024, out_features=1024, bias=True)
        )
        (sa_layer_norm): LayerNorm((1024,), eps=1e-12, elementwise_affine=True)
        (ffn): FFN(
          (dropout): Dropout(p=0.1, inplace=False)
          (lin1): Linear(in_features=1024, out_features=3072, bias=True)
      

### **model_distilprotbert model on Signal-peptide data**

In [26]:
import torch
from torch.utils.data import DataLoader, TensorDataset
import torch.nn as nn
!pip install torchmetrics

import torchmetrics

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 806.1/806.1 kB 5.3 MB/s eta 0:00:00


In [27]:

tokenized = tokenizer(list(signalp_train_data['seq']), max_length=1024, padding=True, truncation=True)
train_inputs = torch.tensor(tokenized['input_ids'])
train_labels=signalp_train_data['label']
label_data = torch.tensor(train_labels)
train_dataset = TensorDataset(train_inputs, label_data )
train_dataloader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)

In [33]:
tokenized = tokenizer(list(signalp_test_data['seq']), max_length=1024, padding=True, truncation=True)
test_inputs = torch.tensor(tokenized['input_ids'])
test_labels=signalp_test_data['label']
label_data = torch.tensor(test_labels)
test_dataset = TensorDataset(test_inputs, label_data )
test_dataloader = DataLoader(test_dataset, batch_size=batch_size, shuffle=True)

In [30]:
import torch
import torch.optim as optim



# Training loop
num_epochs = 10
optimizer = optim.Adam(  model_distilprotbert.parameters(), lr=0.001)
criterion = nn.CrossEntropyLoss()

# **Training The Distealibert model**

In [31]:
linear_layer = nn.Linear(1024, 2)
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model_distilprotbert.to(device)
for epoch in range(1):
    model_distilprotbert.train()
    total_loss = 0.0
    running_loss = 0.0
    correct = 0

    for batch_num, (inputs, labels) in enumerate (train_dataloader):
        inputs, labels = inputs.to(device), labels.to(device)

        optimizer.zero_grad()

        outputs = model_distilprotbert(inputs)

        cls_hidden_state = outputs.last_hidden_state[:, 0, :]  # Assuming [CLS] is at index 0
        output_dim = model_distilprotbert.config.hidden_size
        classification_layer = nn.Linear(output_dim, 2)


        # You can now use cls_hidden_state to calculate the logits

        logits = classification_layer (cls_hidden_state)

        logits = torch.argmax(logits, dim=1).float()

        logits.requires_grad = True
        loss = criterion(logits.float(), labels.float())
        loss.backward()
        optimizer.step()
        correct += (logits.float() == labels.float()).sum()

    accuracy = 100 * correct / len( train_dataloader.dataset)
    loss = running_loss / batch_num
    print(f"Train loss: {loss:.3f}, Train accuracy: {accuracy:.3f}%")


Train loss: 0.000, Train accuracy: 51.048%


# **Test accuracy on Singnal peptide data using the Pretrained Disteli bert model**

In [34]:
accuracy_metric_test = torchmetrics.Accuracy(task='binary')
precision_metric_test = torchmetrics.Precision(num_classes=2, average='binary', task='binary')
recall_metric_test = torchmetrics.Recall(num_classes=2, average='binary', task='binary')




# ... (your previous code)

# Assuming you have a test DataLoader named test_dataloader
for epoch in range(1):
    model_distilprotbert.eval()  # Set the model to evaluation mode
    all_labels_test = []
    all_logits_test = []

    for batch_num, (inputs_test, labels_test) in enumerate(test_dataloader):
        inputs_test, labels_test = inputs_test.to(device), labels_test.to(device)

        with torch.no_grad():
            outputs_test = model_distilprotbert(inputs_test)

        cls_hidden_state_test = outputs_test.last_hidden_state[:, 0, :]
        output_dim_test = model_distilprotbert.config.hidden_size
        classification_layer_test = nn.Linear(output_dim_test, 2)

        logits_test = classification_layer_test(cls_hidden_state_test)

        logits_test = torch.argmax(logits_test, dim=1).float()

        all_labels_test.extend(labels_test.cpu().numpy())
        all_logits_test.extend(logits_test.cpu().numpy())

    # Convert the lists to PyTorch tensors
    all_labels_test = torch.tensor(all_labels_test)
    all_logits_test = torch.tensor(all_logits_test)


    # Calculate precision on test data using torchmetrics
    precision_test = precision_metric_test(all_logits_test, all_labels_test)
    print(f"Test precision: {precision_test:.3f}")
    accuracy_test = accuracy_metric_test(all_logits_test, all_labels_test)

    print(f"Test accuracy: {accuracy_test:.3f}")
    recall = recall_metric_test(all_logits_test, all_labels_test)
    print(f"Test recall: {recall:.3f}")
    # F1 Score
    f1 = 2 * (precision_test * recall) / (precision_test + recall)

    print("F1 Score:", f1.item())





Test precision: 0.155
Test accuracy: 0.513
Test recall: 0.448
F1 Score: 0.23018291592597961


# **Test accuracy on Singnal peptide data without using the Pretrained model**

In [ ]:
# Fill this cell to evaluate the performance of model_finetuned_SP on signal peptide test data
# You can use sklearn library to calculate these metrics.

# accuracy
# precision
# recall
# f1-score

# YOUR CODE HERE
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Assuming you have the true labels (y_true) and predicted labels (y_pred) for your test data
# Replace y_true and y_pred with your actual data

# Calculate accuracy
import pandas as pd

# Assuming your CSV file is named 'your_file.csv'
#file_path = '/signalp_pred.csv'  # Adjust the file path accordingly

# Read the CSV file into a DataFrame
df = pd.read_csv('signalp_pred.csv')

accuracy = accuracy_score(df.label,df.pred_label)

# Calculate precision
precision = precision_score(df.label,df.pred_label)

# Calculate recall
recall = recall_score(df.label,df.pred_label)

# Calculate F1-score
f1 = f1_score(df.label,df.pred_label)

# Print the metrics
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")





Accuracy: 0.1623
Precision: 0.1623
Recall: 1.0000
F1-score: 0.2793


# Distelii bert pretrainned model :  Test precision: 0.156
# Test accuracy: 0.507
# Test recall: 0.464
# F1 Score: 0.23410621285438538

# Without  Distelli bert:
# Accuracy: 0.1623
# Precision: 0.1623
# Recall: 1.0000
# F1-score: 0.2793

In [ ]:
# Put your code here for fine-tuning and evaluating DistilProtBert on whichever task you choose

# YOUR CODE HERE

In [ ]:
# Put your code here for fine-tuning and evaluating the classifer without pretraining on whichever task you choose

# YOUR CODE HERE


In [ ]:
# Write your report here.

<h2>Submission</h2>

<hr style="border-top: 5px solid orange; margin-top: 1px; margin-bottom: 1px"></hr>

<p style="text-align: justify;">You need to submit a Jupyter Notebook (*.ipynb) file that contains your completed code and results.
<p>
Your answers to sections 1-6 will be evaluated based on the correctness of your implementation. Your answer to section 7 will be evaluated based on the correctness of your implementation, the results you report, and the quality of your analysis.

<span>The file name should be in <strong>FirstName_LastName</strong> format</span>.</p>
<p style="text-align: justify;"><span>DO NOT INCLUDE EXTRA FILES, SUCH AS THE INPUT DATASETS</span>, in your submission;</p>
<p style="text-align: justify;">Please download your assignment after submission and make sure it is not corrupted or empty! We will not be responsible for corrupted submissions and will not take a resubmission after the deadline.</p>

Need Help?
If you need help with this assignment, please get in touch with TAs via their emails, or go to their office hours.
You are highly encouraged to ask your question on the designated channel for Assignment o on Microsoft Teams (not necessarily monitored by the instructor/TAs). Feel free to help other students with general questions. However, DO NOT share your solution.<hr style="border-top: 5px solid orange; margin-top: 1px; margin-bottom: 1px"></hr>